File: MyProjects/MAP/Assignment 7/Interpreter/Model/Statements/ReadFileStmt.cs
using System;
using System.Collections.Generic;
using System.Text;
using Interpreter.Model.Expressions;
using System.IO;
using Interpreter.Exceptions;
using System.Linq;

namespace Interpreter.Model.Statements
{
    public class ReadFileStmt:IStmt
    {

        private IExpr id;
        private string var;
        public ReadFileStmt(IExpr e, string s) { id = e; var = s; }

        public ProgramState execute(ProgramState state)
        {
            int val = this.id.evaluate(state.SymbolTable);
            if (state.FileTab.exists(val))
            {
                StreamReader b = state.FileTab.getValue(val);
                string line = b.ReadLine();
                int myValue;
                if (line == null)
                    myValue = 0;
                else
                    myValue = int.Parse(line);
                if (state.SymbolTable.exists(this.var))
                {
                    state.SymbolTable.delete(this.var);
                    state.SymbolTable.setValue(this.var, myValue);
                }
                else
                    state.SymbolTable.setValue(this.var, myValue);

            }
            else
                throw new FileNotFoundException("File not found!");
            return state;
        }

        public override string ToString()
        {
            return "ReadFile(" + id + "," + var + ")";
        }
    }
    
}

File: MyProjects/MAP/Practical Exam/Countdown Latch/src/Model/Statements/ReadFileStmt.java
package Model.Statements;

import Model.Expressions.IExpr;
import Model.ProgramState;
import Model.Utilities.IDictionary;

import java.io.BufferedReader;
import java.io.FileNotFoundException;
import java.io.IOException;

public class ReadFileStmt implements IStmt {
    private IExpr id;
    private String var;
    public ReadFileStmt(IExpr e, String s){id=e; var=s;}

    @Override
    public ProgramState execute(ProgramState state) throws IOException {
        IDictionary<String,Integer> dic=state.getSymTable();
        IDictionary<Integer, Integer> heap=state.getHeap();
        int val=id.evaluate(dic,heap);
        IDictionary<Integer,BufferedReader> filetable=state.getFileTable();
        if(filetable.exists(val)) {
            BufferedReader b = filetable.getValue(val);
            String line=b.readLine();
            int myValue;
            if(line==null)
                myValue=0;
            else
                myValue=Integer.parseInt(line);
            if(dic.exists(var)){
                dic.delete(var);
                dic.setValue(var,myValue);
            }
            else
                dic.setValue(var,myValue);

        }else
            throw new FileNotFoundException("File not found!");
        return null;
    }

    public String toString(){
        return "ReadFile("+id+","+var+")";
    }
}
